File: scripts/algorithms/E/Equal Sum Arrays With Minimum Number of Operations/Equal Sum Arrays With Minimum Number of Operations.java
// Runtime: 6 ms (Top 89.13%) | Memory: 97.8 MB (Top 84.45%)
class Solution {
    public int minOperations(int[] nums1, int[] nums2) {
        int m = nums1.length, n = nums2.length;
        if (m > 6 * n || n > 6 * m) return -1;

        int sum1 = 0, sum2 = 0;
        for (int i : nums1) sum1 += i;
        for (int i : nums2) sum2 += i;

        int diff = sum1 - sum2;
        if (diff == 0) return 0;

        return (diff > 0 ? helper(nums1, nums2, diff)
                         : helper(nums2, nums1, -diff));
    }

    private int helper(int[] nums1, int[] nums2, int diff) {
        // count[i] : frequency of numbers that can reduce the diff by i
        int[] count = new int[6];
        for (int num : nums1) count[num - 1]++;
        for (int num : nums2) count[6 - num]++;

        int res = 0;
        for (int i = 5; i > 0; i--) {
            int c = Math.min(count[i], diff / i + (diff % i == 0 ? 0 : 1));

            res += c;
            diff -= c * i;

            if (diff <= 0) break;
        }
        return res;
    }
}

File: scripts/algorithms/E/Equal Sum Arrays With Minimum Number of Operations/Equal Sum Arrays With Minimum Number of Operations.js
// Runtime: 88 ms (Top 88.89%) | Memory: 52.30 MB (Top 77.78%)

var minOperations = function(nums1, nums2) {
    /*
    Two impossible cases:
    min1...max1 ... min2...max2 -> we can't make them equal
    min2...max2 ... min1...max1 -> we can't make them equal
    */
    let min1 = nums1.length * 1;
    let max1 = nums1.length * 6;
    let min2 = nums2.length * 1;
    let max2 = nums2.length * 6;
    if (min2 > max1 || min1 > max2) {
        return -1;
    }
    let sum1 = nums1.reduce((acc,cur) => acc + cur);
    let sum2 = nums2.reduce((acc,cur) => acc + cur);
    if (sum1 === sum2) return 0;
    if (sum1 < sum2) {
        return helper(nums1, nums2, sum1, sum2);
    } else {
        return helper(nums2, nums1, sum2, sum1);
    }
    // T.C: O(M + N), M = # of nums1, N = # of nums2
    // S.C: O(1)
};

// Condition: sum of A < sum of B
// the idea is to add the maximum possible value to sumA and
// subtract the maximum possible value from sumB so that
// we make sumA >= sumB as soon as possible
function helper(A, B, sumA, sumB) {
    let freqA = new Array(7).fill(0);
    let freqB = new Array(7).fill(0);
    for (let i = 0; i < A.length; i++) {
        freqA[A[i]]++;
    }
    for (let i = 0; i < B.length; i++) {
        freqB[B[i]]++;
    }
    let count = 0;
    for (let i = 1; i <= 6; i++) {
        // increase sumA
        while (freqA[i] > 0 && sumA < sumB) {
            sumA += 6-i;
            freqA[i]--;
            count++;
        }
        let j = 7-i;
        // decrease sumB
        while (freqB[j] > 0 && sumA < sumB) {
            sumB -= j-1;
            freqB[j]--;
            count++;
        }
        if (sumA >= sumB) break;
    }
    return count;
}
